### DISCUSSION TITLE
Typing Websocket

### DISCUSSION THREAD

**Discussion by josejooj** [kartikk221/hyper-express#177] [Q&A]
Hello! I'm working now with websocket, learning about upgrade i did a code like this:

```ts
type CoolTyping = { session: { id: string } };

interface WebSocketRoute<T> {
    pattern: string,
    WSOptions?: HyperExpress.WSRouteOptions
    upgrade: HyperExpress.UserRouteHandler
    handler: HyperExpress.WSRouteHandler<T>
}

const obj: WebsocketRoute<CoolTyping> = {
    pattern: "connect",
    WSOptions: {},
    upgrade: async (req, res) => res.upgrade({ session: { id: "123" } } as CoolTyping)
    handler: async (ws) => console.log(ws.context.session.id, ws.context.session);
}
```
But when i start to work with ws.context i get this error 
```ts
console.log(ws.context.session.id) // Property 'id' does not exist on type 'string' ts(2339) 
```
So i want to know: How can i do to `ws.context` have the type of `CoolTyping`? (I don't want to do `(ws.context as CoolTyping).session.id` every time

OBS: The code runs without problems, i only want to know how to type the context

**Comment by kartikk221** [comment]
Hello, so the thing is the `ws.context` is typed very loosely. Since hyper-express does not support a generic based specification for Websockets due to them having the ability to be loosely coupled across `upgrade` and `ws` routes. In the future a generic specification is planned. For now, you could do something like this:
```js
interface MyWebsocketContext {
    id: string;
    created_at: number;
};

server.ws('/something', (ws) => {
    const context = ws.context as MyWebsocketContext;
});
```
